### PR TITLE
refactor: remove redundant `ImplicitDeallocate` node addition during AST -> ASR transformation

### LINKED ISSUE #2763: Duplicate deallocate in the ASR
```fortran
program main
integer, allocatable :: d(:)
allocate(d(3))
end program
```
ASR after passes:
```clojure
...
(ImplicitDeallocate
    [(Var 2 d)]
)
(ImplicitDeallocate
    [(Var 2 d)]
)
...
```

The ImplicitDeallocate nodes are inserted in `array_op` and `insert_deallocate` passes. I think we have to remove this in `array_op`.

### PR DESCRIPTION
**Pull request by kmr-srbh**
Cleanup as part of #3770. Fixes #2763.

We handle this explicitly in the `insert_deallocate` pass, so this is not required here.

### CODE REVIEW

File: tests/reference/asr-allocate_01-f3446f6.stdout [hunk 1]
```diff
@@ -1633,11 +1630,6 @@
                             ()
                         )]
                         []
-                    )
-                    (ImplicitDeallocate
-                        [(Var 2 a)
-                        (Var 2 b)
-                        (Var 2 c)]
```
Comment (kmr-srbh) [inline tests/reference/asr-allocate_01-f3446f6.stdout:1640]: We do not remove this, just that instead of being added here after AST -> ASR transformation, we add this in the `insert_deallocate` pass.

**Comment by certik** [comment]
We have to be very careful here.

First, the `insert_deallocate` doesn't seem to insert the deallocate correctly:
```c++
        void visit_Symbol(const T& x) {
            Vec<ASR::expr_t*> to_be_deallocated;
            to_be_deallocated.reserve(al, 1);
            for( auto& itr: x.m_symtab->get_scope() ) {
                if( ASR::is_a<ASR::Variable_t>(*itr.second) &&
                    ASR::is_a<ASR::Allocatable_t>(*ASRUtils::symbol_type(itr.second)) &&
                    ASRUtils::is_array(ASRUtils::symbol_type(itr.second)) &&
                    ASRUtils::symbol_intent(itr.second) == ASRUtils::intent_local ) {
                    to_be_deallocated.push_back(al, ASRUtils::EXPR(
                        ASR::make_Var_t(al, x.base.base.loc, itr.second)));
                }
            }
            if( to_be_deallocated.size() > 0 ) {
                T& xx = const_cast<T&>(x);
                Vec<ASR::stmt_t*> body;
                body.from_pointer_n_copy(al, xx.m_body, xx.n_body);
                body.push_back(al, ASRUtils::STMT(ASR::make_ImplicitDeallocate_t(
                    al, x.base.base.loc, to_be_deallocated.p, to_be_deallocated.size())));
                xx.m_body = body.p;
                xx.n_body = body.size();
            }
        }
```
This code only appends it the given body. However, one must actually properly insert this into every place that can exit the function (if statements, etc.).

Second, we really want to enforce this using ASR verify, to check that all things are properly deallocated and that leaks cannot happen. However that means we need to do this when constructing the ASR. Alternatively, we need to create a pass and run it just to fix this. However for compilation speed reasons, we should do it in one pass.

We can have a helper pass to use for debugging or to get things done quickly, but it seems like a wrong design to have a pass that one has to call every time to "fix up the ASR".

An alternative design would be that deallocation is not enforced via ASR verify, nor required. In that case, the semantics would be that things get deallocated automatically. Then we could have a pass that inserts the deallocation nodes correctly. It might not be a bad design, since the deallocation logic would be concentrated in a pass. But we need to design this well.

**Comment by assem2002** [comment]
If we go with `insert_deallocate` pass, This PR seems to be a good step towards handling the finalization issue.
If we go with ASR_verify(), So we gotta keep that for sure.

**Comment by certik** [comment]
Keep the PR open for now. We'll figure out a design very soon.

File: tests/reference/llvm-allocate_03-495d621.stdout [hunk 2]
```diff
@@ -288,32 +288,9 @@ else13:                                           ; preds = %ifcont11
...
-  %176 = load i8*, i8** %174, align 8
-  call void @_lfortran_free(i8* %176)
-  %177 = getelementptr %array, %array* %167, i32 0, i32 0
-  store i32* null, i32** %177, align 8
-  br label %ifcont17
-
-else16:                                           ; preds = %ifcont14
-  br label %ifcont17
-
-ifcont17:                                         ; preds = %else16, %then15
```
Comment (certik) [inline tests/reference/llvm-allocate_03-495d621.stdout:313]: Why were these lines removed?

Comment (kmr-srbh) [reply tests/reference/llvm-allocate_03-495d621.stdout:313]: This is correct actually, the code was redundant. Before this PR, we were adding the `ImplicitDeallocate` node twice, first in AST -> ASR, then in the pass. In this PR, I removed exactly the same logic from AST -> ASR as had been implemented in the pass, so we only remove redundancy, Currently, the case of `SubroutineCall` is still handled during AST -> ASR. In the next PR, I will port those changes here.

Talking of redundancy, we need to take care of the user allocatables being deallocated twice, first through `ExplicitDeallocate` node inserted during AST -> ASR, then through `ImplicitDeallocate` inserted in this pass. As `ImplicitDeallocate` does not throw an error for deallocating unallocated `allocatable` we do not get an error, but this still adds redundancy.

```fortran
program main
    implicit none
        integer, allocatable :: c(:)
        allocate(c(3))
        deallocate(c)
end program
```
```clojure
;; ASR after applying the pass: insert_deallocate
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        2
                        {
                            c:
                                (Variable
                                    2
                                    c
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Allocatable
                                        (Array
                                            (Integer 4)
                                            [(()
                                            ())]
                                            DescriptorArray
                                        )
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    main
                    []
                    [(Allocate
                        [((Var 2 c)
                        [((IntegerConstant 1 (Integer 4) Decimal)
                        (IntegerConstant 3 (Integer 4) Decimal))]
                        ()
                        ())]
                        ()
                        ()
                        ()
                    )
                    (ExplicitDeallocate
                        [(Var 2 c)]
                    )
                    (ImplicitDeallocate
                        [(Var 2 c)]
                    )]
                )
        })
    []
)
```

**Comment by certik** [comment]
Ok. Go ahead and merge this.
